contracts/math/uints.ral

```diff
@@ -90,5 +90,76 @@ Abstract Contract U512Impl () {
                 lower: deltaL
             }
         }
+
+    pub fn bigAdd512(a: U512, b: U512) -> U512 {
+        let (r0, overflow) = overflowingAdd(a.lower, b.lower)
+        let r1 = wrappingAdd(wrappingAdd(a.higher, b.higher), overflow)
+        return U512 { higher: r1, lower: r0 }
+    }
+
+    pub fn bigMul(a: U256, b: U256, denominator: U256) -> U512 {
+        let aLower = low128(a)
+        let aHigher = high128(a)
+        let bLower = low128(b)
+        let bHigher = high128(b)
+
+        let rLowerLower = aLower * bLower
+        let rHigherHigher = aHigher * bHigher
+        let rHigherLower = aHigher * bLower
+        let rLowerHigher = aLower * bHigher
+
+        let mut result = bigAdd512(U512 { higher: rHigherHigher, lower: rLowerLower }, U512 { higher: high128(rHigherLower), lower: low128(rHigherLower) << 128 })
+        result = bigAdd512(result, U512 { higher: high128(rLowerHigher), lower: low128(rLowerHigher) << 128 })
+
+        result = bigDiv(result, denominator, 1)
+
+        return result
+    }
+
+    pub fn bigMulUp(a: U256, b: U256, denominator: U256) -> U512 {
+        let a0 = low128(a)
+        let a1 = high128(a)
+        let b0 = low128(b)
+        let b1 = high128(b)
+
+        let r0 = a0 * b0
+        let r1 = a1 * b1
+        let r2 = a1 * b0
+        let r3 = a0 * b1
+
+        let mut result = bigAdd512(U512 { higher: r1, lower: r0 }, U512 { higher: high128(r2), lower: low128(r2) << 128 })
+        result = bigAdd512(result, U512 { higher: high128(r3), lower: low128(r3) << 128 })
+
+        result = bigAdd512(result, U512 { higher: 0, lower: 10 ** denominator - 1 })
+        result = bigDiv(result, denominator, 1)
+
+        return result
+    }
+    
+    pub fn overflowingAdd(a: U256, b: U256) -> (U256, U256) {
+        if (u256Max!() - a >= b) {
+            return a + b, 0
+        } else {
+            return b + (u256Max!() - a) - 1, 1
+        }
+    }
+
+    pub fn wrappingAdd(a: U256, b: U256) -> U256 {
+        let u256Max = u256Max!()
```

Comment:
You do not need to define additional variable.
It is already defined
`const MaxU256 = 115792089237316195423570985008687907853269984665640564039457584007913129639935`